lib/service/kubernetes.go

```diff
@@ -0,0 +1,241 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tlsConfig, err := conn.ServerIdentity.TLSConfig(cfg.CipherSuites)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	kubeServer, err := kubeproxy.NewTLSServer(kubeproxy.TLSServerConfig{
+		ForwarderConfig: kubeproxy.ForwarderConfig{
+			Namespace:      defaults.Namespace,
+			Keygen:         cfg.Keygen,
+			ClusterName:    conn.ServerIdentity.Cert.Extensions[utils.CertExtensionAuthority],
+			Auth:           authorizer,
+			Client:         conn.Client,
+			DataDir:        cfg.DataDir,
+			AccessPoint:    accessPoint,
+			ServerID:       cfg.HostUUID,
+			KubeconfigPath: cfg.Kube.KubeconfigPath,
+		},
+		TLS:           tlsConfig,
+		AccessPoint:   accessPoint,
+		Component:     teleport.ComponentKube,
+		LimiterConfig: cfg.Kube.Limiter,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if retErr != nil {
+			warnOnErr(kubeServer.Close())
+		}
+	}()
+	process.RegisterCriticalFunc("kube.serve", func() error {
+		if conn.UseTunnel() {
+			log.Info("Starting Kube service via proxy reverse tunnel.")
+			utils.Consolef(cfg.Console, teleport.ComponentKube, "Kubernetes service %s:%s is starting via proxy reverse tunnel.", teleport.Version, teleport.Gitref)
+		} else {
+			log.Infof("Starting Kube service on %v.", listener.Addr())
+			utils.Consolef(cfg.Console, teleport.ComponentKube, "Kubernetes service %s:%s is starting on %v.", teleport.Version, teleport.Gitref, listener.Addr())
+		}
+		err := kubeServer.Serve(listener)
+		if err != nil {
+			if err == http.ErrServerClosed {
+				return nil
+			}
+			return trace.Wrap(err)
+		}
+		return nil
+	})
+
+	// Start the heartbeat to announce kubernetes_service presence.
+	heartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
+		Mode:            srv.HeartbeatModeKube,
+		Context:         process.ExitContext(),
+		Component:       teleport.ComponentKube,
+		Announcer:       conn.Client,
+		GetServerInfo:   kubeServer.GetServerInfo,
+		KeepAlivePeriod: defaults.ServerKeepAliveTTL,
+		AnnouncePeriod:  defaults.ServerAnnounceTTL/2 + utils.RandomDuration(defaults.ServerAnnounceTTL/10),
+		ServerTTL:       defaults.ServerAnnounceTTL,
+		CheckPeriod:     defaults.HeartbeatCheckPeriod,
+		Clock:           cfg.Clock,
+		OnHeartbeat: func(err error) {
+			if err != nil {
+				process.BroadcastEvent(Event{Name: TeleportDegradedEvent, Payload: teleport.ComponentKube})
+			} else {
+				process.BroadcastEvent(Event{Name: TeleportOKEvent, Payload: teleport.ComponentKube})
+			}
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	process.RegisterCriticalFunc("kube.heartbeat", heartbeat.Run)
+
+	// Cleanup, when process is exiting.
+	process.onExit("kube.shutdown", func(payload interface{}) {
+		// Clean up items in reverse order from their initialization.
+		warnOnErr(heartbeat.Close())
+		if payload != nil {
+			// Graceful shutdown.
+			warnOnErr(kubeServer.Shutdown(payloadContext(payload)))
+			agentPool.Stop()
```

Comment:
@awly Just noticed something in this old PR as I'm working on a similar part in db access. Looks like in case of direct dial (`!conn.UseTunnel()` case above), agentPool will stay uninitialized so these stop/wait calls will probably panic in that case?

Reply:
I made stop/wait nil-safe: https://github.com/gravitational/teleport/pull/4611/files#diff-b9f7168b21e0fd7edd940ce7547de3171339012b8d4064d135e50340c6790c9fR171-R185